src/components/Card/TodosCard/TodosCard.tsx

```diff
@@ -0,0 +1,272 @@
+  const [todos, setTodos] = useState<Array<Todo>>([]);
+  const [todosFilter, setTodosFilter] = useState<string>(Filter.ALL);
+  const [todoToUpdate, setTodoToUpdate] = useState<Todo>();
+  const [expiredTodo, setExpiredTodo] = useState<Todo>();
+  const [isTimeModalOpen, setTimeModalState] = useState(false);
+  const [isDeadlineModalOpen, setDeadlineModalState] = useState(false);
+  const remaindersWorker = useMemo(
+    () => new Worker("/workers/RemaindersWorker.js"),
+    []
+  );
+
+  remaindersWorker.onmessage = (event: MessageEvent) => {
+    setExpiredTodo(todos.find(todo => todo.id === event.data));
+    setDeadlineModalState(true);
+  };
+
+  // TODO: please think about what could happen if the component will unmount before the response is returned from the server
+  // implement this solution (important)
+  // I researched and found three solutions to this problem:
+  // 1. Managing an isMounted property, that will be set to true when the component is mounted
+  // and will be set false in the useEffect return function, https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html - this article
+  // says that the deprecated isMounted function is an anti pattern and not recommended to use and they offer managing the isMounted property.
+  // 2. Another solution offered in this article is creating a warper for all promises, which makes the promise cancelable.
+  // 3. Third option is to move the todos array to Redux and then replacing the setState functions with dispatch.
+  // I'm not really sure what the best approach is, and would like your opinion.
+  // EDIT: I decided to go with managing my own isMounted after seeing Dan Abramov recommends this method.
+  // I actually found another option and it's using axios cancellation token.
+  useEffect(() => {
+    let isMounted = true;
+    const fetchTodos = async () => {
+      try {
+        const todos = (await TodoService.getTodos()).data;
+        if (isMounted) {
+          setTodos(todos);
+          remaindersWorker.postMessage({ type: "add", data: todos });
+        }
+      } catch (error) {
+        fireSwalError("The todos couldn't be fetched from the server");
+      }
+    };
+
+    fetchTodos();
+
+    return () => {
+      isMounted = false;
+    };
+  }, [remaindersWorker]);
+
+  const getNumberOfUncompletedTodos = () => {
+    // TODO: remove unused parentheses
+    // ANSWER: It's because of my prettier config.. I changed it
+    return todos.filter(todo => !todo.isCompleted).length;
+  };
+
+  const clearAllCompleted = useCallback(async (): Promise<void> => {
+    const [completedTodos, unCompletedTodos] = todos.reduce<Todo[][]>(
+      (accumulator, todo) => {
+        accumulator[todo.isCompleted ? 0 : 1].push(todo);
+        return accumulator;
+      },
+      [[], []]
+    );
+
+    // TODO: that's not efficient, you make http requests in loop,
+    // what will happen if there are alot of todos?
+    // ANSWER: The http requests aren't made in a loop
```

Comment:
please refer to here: https://stackoverflow.com/questions/40605427/how-to-not-run-a-promise-before-i-ask-to

the promise constructor is immediately invoked and therefore your delete.

anyway, even if you did use promises, you create a promise array of the same length as your todos, and then you call the same amount of requests, even if you don't do it in map function.

the same result would be to run at foreach and not wait for every delete operation.

read further about it and write down your conclusions

Reply:
I wrote it because I tried to debug it, and I couldn't see the http requests being made in the Network tab,
I'll to read more and get back to you.